qiling/os/dos/dos.py

```diff
@@ -0,0 +1,152 @@
+                    self.ql.reg.ax = self.handle_next
+                    self.handle_next += 1
+                    self.clear_cf()
+                elif ah == 0x3e:
+                    hd = self.ql.reg.bx
+                    if hd not in self.dos_handles:
+                        self.ql.reg.ax = 0x6
+                        self.set_cf()
+                    else:
+                        f = self.dos_handles[hd]
+                        f.close()
+                        del self.dos_handles[hd]
+                        self.clear_cf()
+                elif ah == 0x3f:
+                    hd = self.ql.reg.bx
+                    if hd not in self.dos_handles:
+                        self.ql.reg.ax = 0x6
+                        self.set_cf()
+                    else:
+                        f = self.dos_handles[hd]
+                        buffer = self.calculate_address(self.ql.reg.ds, self.ql.reg.dx)
+                        sz = self.ql.reg.cx
+                        rd = f.read(sz)
+                        ql.mem.write(buffer, rd)
+                        self.clear_cf()
+                        self.ql.reg.ax = len(rd)
+                elif ah == 0x40:
+                    hd = self.ql.reg.bx
+                    if hd not in self.dos_handles:
+                        self.ql.reg.ax = 0x6
+                        self.set_cf()
+                    else:
+                        f = self.dos_handles[hd]
+                        buffer = self.calculate_address(self.ql.reg.ds, self.ql.reg.dx)
+                        sz = self.ql.reg.cx
+                        rd = self.ql.mem.read(buffer, sz)
+                        f.write(bytes(rd))
+                        self.clear_cf()
+                        self.ql.reg.ax = len(rd)
+                elif ah == 0x41:
+                    fname = self.read_dos_string_from_ds_dx()
+                    real_path = PathUtils.convert_for_native_os(self.ql.rootfs, self.ql.cur_path, fname)
+                    try:
+                        os.remove(real_path)
+                        self.clear_cf()
+                    except OSError:
+                        self.ql.reg.ax = 0x5
+                        self.set_cf()
+                elif ah == 0x43:
+                    self.ql.reg.cx = 0xFFFF
+                    self.clear_cf()
+                else:
+                    raise NotImplementedError()
+            else:
+                raise NotImplementedError()
+        self.ql.hook_intr(cb)
+
+    def run(self):
+        if self.ql.exit_point is not None:
+            self.exit_point = self.ql.exit_point
+
+        if  self.ql.entry_point is not None:
+            self.ql.loader.elf_entry = self.ql.entry_point
+        else:
+            self.ql.entry_point = self.ql.loader.start_address
+        if not self.ql.shellcoder:
+            try:
+                self.ql.reg.write("ds", self.ql.loader.cs)
+                self.ql.reg.write("es", self.ql.loader.cs)
```

Comment:
all the setup code should not be in run.
most of our setup code in loader

Reply:
See commit 80f568e.